planner/core/hashcode.go

```diff
@@ -97,3 +81,305 @@ func (p *LogicalLimit) HashCode() []byte {
+	// PlanType + SelectOffset + Encode(WindowFuncDescs) + Encode(PartitionBys) + Encode(OrderBys) + Frame
+	// WindowFuncDescs are commonly has at most one arg, so we pre-alloc 30 bytes for each WindowFuncDesc's hashcode.
+	// we pre-alloc total bytes size = SizeOf(PlanType)+SizeOf(SelectOffset)+SizeOf(Encode(WindowFuncDescs))+SizeOf(Encode(PartitionBys))+SizeOf(Encode(OrderBys))+SizeOf(Frame)
+	//								 = 4+4+(4+len(WindowFuncDescs)*(4+Sizeof(WindowFuncDesc)))+(4+len(PartitionBys)*(4+Sizeof(PartitionBy)))+(4+len(OrderBys)*(4+Sizeof(OrderBy)))+SizeOf(Frame.hashcode)
+	//								 = 20+34*len(WindowFuncDescs)+14*(len(p.PartitionBy)+len(p.OrderBy))+SizeOf(Frame.hashcode)
+	frameHashcode := p.Frame.HashCode(p.ctx.GetSessionVars().StmtCtx)
+
+	result := make([]byte, 0, 20+len(p.WindowFuncDescs)*34+14*(len(p.PartitionBy)+len(p.OrderBy))+len(frameHashcode))
+	result = codec.EncodeIntAsUint32(result, plancodec.TypeStringToPhysicalID(p.tp))
+	result = codec.EncodeIntAsUint32(result, p.SelectBlockOffset())
+
+	wfdHashCode := func(i int) []byte { return p.WindowFuncDescs[i].HashCode(p.ctx.GetSessionVars().StmtCtx) }
+	result = codec.Encode(result, wfdHashCode, len(p.WindowFuncDescs))
+
+	partitionByHashCode := func(i int) []byte { return p.PartitionBy[i].HashCode(p.ctx.GetSessionVars().StmtCtx) }
+	result = codec.EncodeAndSort(result, partitionByHashCode, len(p.PartitionBy))
+
+	orderByHashCode := func(i int) []byte { return p.OrderBy[i].HashCode(p.ctx.GetSessionVars().StmtCtx) }
+	result = codec.Encode(result, orderByHashCode, len(p.OrderBy))
+
+	result = append(result, frameHashcode...)
+
+	return result
+}
+
+// HashCode implements LogicalPlan interface.
+func (p *LogicalMemTable) HashCode() []byte {
+	// PlanType + SelectOffset + TableInfo.ID + QueryTimeRange
+	result := make([]byte, 0, 32)
+	result = codec.EncodeIntAsUint32(result, plancodec.TypeStringToPhysicalID(p.tp))
+	result = codec.EncodeIntAsUint32(result, p.SelectBlockOffset())
+
+	result = codec.EncodeInt(result, p.TableInfo.ID)
+
+	result = codec.EncodeInt(result, p.QueryTimeRange.From.Unix())
+	result = codec.EncodeInt(result, p.QueryTimeRange.To.Unix())
+	return result
+}
+
+// HashCode implements LogicalPlan interface.
+func (p *DataSource) HashCode() []byte {
+	// PlanType + SelectOffset + tableInfo.ID + Encode(allConds)
+	// allConds are commonly `ScalarFunction`s, whose hashcode usually has a
+	// length larger than 20, so we pre-alloc 25 bytes for each expr's hashcode.
+	// we pre-alloc total bytes size = SizeOf(PlanType)+SizeOf(SelectOffset)+SizeOf(tableInfo.ID)+SizeOf(Encode(allConds))
+	//								 = 4+4+8+(4+len(allConds)*(4+Sizeof(condition.hashcode)))
+	//								 = 20+len(allConds)*29
+	if p == nil {
+		return nil
+	}
+
+	result := make([]byte, 0, 20+len(p.allConds)*29)
+	result = codec.EncodeIntAsUint32(result, plancodec.TypeStringToPhysicalID(p.tp))
+	result = codec.EncodeIntAsUint32(result, p.SelectBlockOffset())
+	result = codec.EncodeInt(result, p.tableInfo.ID)
```

Comment:
I think DataSource's hash code should use `baseLogicalPlan.HashCode()`. Use `tableInfo` is not right in the self-join senario. For example, `select * from t t1, t t2 where ....` , t1 and t2 have the same tableInfo.ID`.

Reply:
`t1` and `t2` is in different groups.

and can we ignore the difference between 
```
		join
		/  \
	       t1  t2
``` 
and 
```
		join
		/  \
	       t2  t1
```
?